File: 백준/Gold/2461. 대표 선수/대표 선수.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());

        int[][] school = new int[n][m];
        int[] indexes = new int[n];
        PriorityQueue<int[]> pq = new PriorityQueue<>(Comparator.comparingInt(a -> a[0]));

        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < m; j++) {
                school[i][j] = Integer.parseInt(st.nextToken());
            }
            Arrays.sort(school[i]);
            pq.offer(new int[]{school[i][0], i});
        }

        int maxValue = Arrays.stream(school).mapToInt(row -> row[0]).max().orElse(0);
        int result = Integer.MAX_VALUE;

        while (!pq.isEmpty()) {
            int[] current = pq.poll();
            int value = current[0];
            int classIndex = current[1];

            result = Math.min(result, maxValue - value);

            indexes[classIndex]++;
            int nextIndex = indexes[classIndex];
            if (nextIndex >= m) {
                break;
            }

            maxValue = Math.max(maxValue, school[classIndex][nextIndex]);
            pq.offer(new int[]{school[classIndex][nextIndex], classIndex});
        }

        System.out.println(result);
    }
}

File: 백준/Gold/2461. 대표 선수/대표 선수.py
from heapq import heappop, heappush
import sys
input = lambda: sys.stdin.readline().strip()

n,m = map(int,input().split())

school = []
for i in range(n):
    school.append(sorted(map(int,input().split())))
    
indexes = [0]*n

q = []
max_value = school[0][0]
for i in range(n):
    max_value = max(max_value, school[i][0])
    heappush(q, [school[i][0],i])
    
result = float('inf')
while q:
    v,th = heappop(q)
    result=min(result,max_value-v)
    indexes[th]+=1
    index = indexes[th]
    if index>=m:
        break
    # print(th,index)
    max_value = max(max_value, school[th][index])
    heappush(q, [school[th][index],th])
    
print(result)